File: cpp/src/edu/uf/intracellularState/NeutrophilStateModel.cpp
/*
 * NeutrophilStateModel.cpp
 *
 *  Created on: Jul 22, 2024
 *      Author: henriquedeassis
 */
#include "IntracellularModel.h"
#include "AspergillusIntracellularModel.h"
#include "../compartments/GridFactory.h"
#include "../interactable/Cell.h"
#include "../interactable/InfectiousAgent.h"
#include "../interactable/Lactoferrin.h"
#include "../interactable/Neutrophil.h"
#include "../compartments/Voxel.h"
#include "../interactable/Cell.h"
#include "../interactable/afumigatus/Afumigatus.h"
#include "../utils/Constants.h"
#include "../utils/Rand.h"
#include <vector>
#include <algorithm>
#include <random>
#include "NeutrophilStateModel.h"
#include <iostream>

namespace edu {
namespace uf {
namespace intracellularState {

using namespace compartments;
using namespace utils;
using namespace afumigatus;

const string NeutrophilStateModel::name = "NeutrophilStateModel";

const int NeutrophilStateModel::ACTIVE = 0;
const int NeutrophilStateModel::DYING = 1;
const int NeutrophilStateModel::APOPTOTIC = 2;
const int NeutrophilStateModel::NETOTIC = 3;

NeutrophilStateModel::NeutrophilStateModel() {
    booleanNetwork = new int[size]();
}

void NeutrophilStateModel::processBooleanNetwork() {
    if (booleanNetwork[APOPTOTIC] > 0 || booleanNetwork[NETOTIC] == 1) {
        return;
    }

    int k = 0;
    vector<int> array(size);
    iota(array.begin(), array.end(), 0);

    while (true) {
        if (k++ > 3) break;
        shuffle(array.begin(), array.end(), Rand::getRand().getGenerator());
        for (int i : array) {
            switch (i) {
                case ACTIVE:
                    booleanNetwork[ACTIVE] = getInput(Afumigatus::DEF_OBJ) |
                                             //getInput(TLRBinder::getBinder()) |
                                             booleanNetwork[ACTIVE];
                    break;
                case DYING:
                    booleanNetwork[DYING] = booleanNetwork[ACTIVE] > 0 ?
                                            (Rand::getRand().randunif() < utils::constexprants::NEUTROPHIL_HALF_LIFE ? 1 : 0) :
                                            0;
                    break;
                case APOPTOTIC:
                    if (booleanNetwork[APOPTOTIC] == 0) {
                        int j = booleanNetwork[DYING] > 0 ?
                                (Rand::getRand().randunif() < utils::constexprants::PR_NEUT_APOPT ? 2 : 1) :
                                0;

                        booleanNetwork[APOPTOTIC] = j == 2 ? 1 : 0;
                        booleanNetwork[NETOTIC] = j == 1 ? 1 : 0;
                    }
                    break;
                case NETOTIC:
                    booleanNetwork[APOPTOTIC] = booleanNetwork[ACTIVE] == 0 ?
                                                (Rand::getRand().randunif() < utils::constexprants::NEUTROPHIL_HALF_LIFE ? 1 : 0) :
                                                0;
                    break;
                default:
                    cerr << "No such interaction " << i << "!" << endl;;
                    break;
            }
        }
    }

    inputs.clear();
    clearPhenotype();
    computePhenotype();
}

void NeutrophilStateModel::computePhenotype() {
    if (booleanNetwork[ACTIVE] > 0) {
        getPhenotype()[ACTIVE] = booleanNetwork[ACTIVE];
        getPhenotype()[Lactoferrin::getMolecule()->getPhenotype()] = booleanNetwork[ACTIVE];
    }
    if (booleanNetwork[APOPTOTIC] == 1) {
        getPhenotype()[APOPTOTIC] = booleanNetwork[APOPTOTIC];
    }
    if (booleanNetwork[NETOTIC] == 1) {
        getPhenotype()[NETOTIC] = booleanNetwork[NETOTIC];
    }
}

void NeutrophilStateModel::updateStatus(int id, int x, int y, int z) {
    Neutrophil* net = dynamic_cast<Neutrophil*>(Cell::get(id));
    if (this->getState(IntracellularModel::LIFE_STATUS) == IntracellularModel::DEAD) {
        return;
    }

    if (net->getBooleanNetwork()->hasPhenotype(APOPTOTIC)) {
        net->die();
        for (Cell* entry : net->getPhagosome()) {
            entry->getBooleanNetwork()->setState(AspergillusIntracellularModel::LOCATION, AspergillusIntracellularModel::RELEASING);
        }
    }
    if (net->getBooleanNetwork()->hasPhenotype(NETOTIC)) {
        GridFactory::getGrid()[x][y][z]->setExternalState(1);
        if (Rand::getRand().randunif() < net->getNetHalfLife()) {
            net->die();
            GridFactory::getGrid()[x][y][z]->setExternalState(0);
            for (Cell* entry : net->getPhagosome()) {
                entry->getBooleanNetwork()->setState(AspergillusIntracellularModel::LOCATION, AspergillusIntracellularModel::RELEASING);
            }
        }
    }

    net->setMaxMoveStep(-1);
    net->setEngaged(false);
}

} // namespace intracellularState
} // namespace uf
} // namespace edu

File: java/src/edu/uf/intracellularState/NeutrophilStateModel.java
package edu.uf.intracellularState;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import edu.uf.compartments.GridFactory;
import edu.uf.interactable.Cell;
import edu.uf.interactable.InfectiousAgent;
import edu.uf.interactable.Lactoferrin;
import edu.uf.interactable.Neutrophil;
import edu.uf.interactable.TLRBinder;
import edu.uf.interactable.Afumigatus.Afumigatus;
import edu.uf.utils.Constants;
import edu.uf.utils.Rand;

public class NeutrophilStateModel extends IntracellularModel{
	
	public static final String name = "NeutrophilStateModel";

	public static final int size = 4;
	//public static final int NUM_RECEPTORS = 4;
	
	public static final int ACTIVE = 0;
	public static final int DYING = 1;
	public static final int APOPTOTIC = 2;
	public static final int NETOTIC = 3;
	
	{
		this.booleanNetwork = new int[size];
	}
	
	
	@Override
	public void processBooleanNetwork(int... args) {
		
		if(this.booleanNetwork[APOPTOTIC] > 0 || this.booleanNetwork[NETOTIC] == 1) {
			//System.out.println(this.booleanNetwork[APOPTOTIC] + " " + this.booleanNetwork[NETOTIC] + " " + this.getPhenotype().get((NeutrophilStateModel.NETOTIC)));
			return;
		}

		int k = 0;
		List<Integer> array = new ArrayList<>(size);
		for(int i = 0; i < size; i++)
			array.add(i);
		while(true) {
			if(k++ > 3)break;
			Collections.shuffle(array, new Random());
			for(int i : array) {
				switch(i) {
					case 0:
						this.booleanNetwork[ACTIVE] = getInput(Afumigatus.DEF_OBJ) | getInput(TLRBinder.getBinder()) | this.booleanNetwork[ACTIVE];
						break;
					case 1:
						
						this.booleanNetwork[DYING] = this.booleanNetwork[ACTIVE] > 0 ? (Rand.getRand().randunif() < Constants.NEUTROPHIL_HALF_LIFE ? 1 : 0) : 0;
						break;
					case 2:
						if(this.booleanNetwork[APOPTOTIC] == 0) {
							int j = this.booleanNetwork[DYING] > 0 ? (Rand.getRand().randunif() < Constants.PR_NEUT_APOPT ? 2 : 1) : 0;
						
							this.booleanNetwork[APOPTOTIC] =  j == 2 ? 1 : 0;
							this.booleanNetwork[NETOTIC] = j == 1 ? 1 : 0; //
							//if(this.booleanNetwork[DYING] > 0) System.out.println(this.booleanNetwork[NETOTIC]  + " " + this.booleanNetwork[APOPTOTIC]);
						}
						break;
					case 3:
						this.booleanNetwork[APOPTOTIC] = this.booleanNetwork[ACTIVE] == 0 ? (Rand.getRand().randunif() < Constants.NEUTROPHIL_HALF_LIFE ? 1 : 0) : 0;
						//System.out.println(this.booleanNetwork[APOPTOTIC]);
						break;
					default:
						System.err.println("No such interaction " + i + "!");
						break;
				}
			}
		}
		
		this.inputs.clear();
		
		this.clearPhenotype();
		
		this.computePhenotype();
		
	}
	
	protected void computePhenotype() {
		if(this.booleanNetwork[ACTIVE] > 0) {
			this.getPhenotype().put(NeutrophilStateModel.ACTIVE, this.booleanNetwork[ACTIVE]);
			this.getPhenotype().put(Lactoferrin.getMolecule().getPhenotype(), this.booleanNetwork[ACTIVE]);
		} 
		if(this.booleanNetwork[APOPTOTIC] == 1) {
			this.getPhenotype().put(NeutrophilStateModel.APOPTOTIC, this.booleanNetwork[APOPTOTIC]);
		}
		if(this.booleanNetwork[NETOTIC] == 1) {
			this.getPhenotype().put(NeutrophilStateModel.NETOTIC, this.booleanNetwork[NETOTIC]);
		}
	}

	@Override
	public void updateStatus(int id, int x, int y, int z) {
		Neutrophil net = (Neutrophil) Cell.get(id);
		if(this.getState(IntracellularModel.LIFE_STATUS) == IntracellularModel.DEAD)
            return;
        
        
        if(net.getBooleanNetwork().hasPhenotype(NeutrophilStateModel.APOPTOTIC)) {
        	net.die();
            for(InfectiousAgent entry : net.getPhagosome())
            	entry.getBooleanNetwork().setState(AspergillusIntracellularModel.LOCATION, AspergillusIntracellularModel.RELEASING);
        }
        if(net.getBooleanNetwork().hasPhenotype(NeutrophilStateModel.NETOTIC)) {
        	GridFactory.getGrid()[x][y][z].setExternalState(1);
        	/*if(this.clock.getCount() >= 48) {
        		this.netHalfLife = Constants.NET_HALF_LIFE*Constants.NET_COUNTER_INHIBITION;
        		System.out.println(this.netHalfLife + " " + Constants.NET_HALF_LIFE);
        	}*/
        	if(Rand.getRand().randunif() < net.getNetHalfLife()) {
        		net.die();
        		GridFactory.getGrid()[x][y][z].setExternalState(0);
        		for(InfectiousAgent entry : net.getPhagosome())
        			entry.getBooleanNetwork().setState(AspergillusIntracellularModel.LOCATION, AspergillusIntracellularModel.RELEASING);
        	}
        }
        
        
        //else if(Rand.getRand().randunif() < Constants.NEUTROPHIL_HALF_LIFE)
           // this.setStatus(Neutrophil.APOPTOTIC);
        //this.setMoveStep(0);
        net.setMaxMoveStep(-1);
        net.setEngaged(false);
		
	}
	
}
